src/odemis/util/comp.py

```diff
@@ -162,22 +162,16 @@ def generate_zlevels(focuser: "Actuator", zrange: Tuple[float, float], zstep: fl
     if n > MAX_ZLEVELS:
         raise IndexError(f"The number of zlevels, {n}, is too large. Reduce the zstep value to < {MAX_ZLEVELS}.")
 
-    # zstep often doesn't exactly fit to have a round number of zlevels. So try
-    # the floor and ceil values for the number of zlevels, and take the one that
-    # give a step closest to the requested zstep.
+    # When the zstep is bigger than the z range
+    # we return the z range itself
     n_f = math.floor(n)
-    n_c = math.ceil(n)
-
-    if n_f <= 1:  # would cause a division by zero
-        # => do the minimum amount of steps
-        n = 2
-    else:
-        # errors
-        ef = abs((zrange_abs[1] - zrange_abs[0]) / (n_f - 1) - abs(zstep))
-        ec = abs((zrange_abs[1] - zrange_abs[0]) / (n_c - 1) - abs(zstep))
-        n = n_f if ef < ec else n_c
-
-    if zstep > 0:
-        return numpy.linspace(zrange_abs[0], zrange_abs[1], n).tolist()
-    elif zstep < 0:
-        return numpy.linspace(zrange_abs[1], zrange_abs[0], n).tolist()
+    if n_f <= 1:
+        return [zrange_abs[0], zrange_abs[1]]
+
+    abs_zstep = abs(zstep)
+    focus_array_min = numpy.arange(focuser_pos, zrange_abs[0] - abs_zstep/10, -abs_zstep)
+    focus_array_max = numpy.arange(focuser_pos, zrange_abs[1] + abs_zstep/10, abs_zstep)
+    zlevels_list = numpy.unique(numpy.append(focus_array_min, focus_array_max)).tolist()
```

Comment:
Why do you need `numpy.unique` here? If it is just for the focuser_pos, you could also write it something like this:
```suggestion
    zlevels_list = numpy.append(focus_array_min[1:][::-1], focus_array_max).tolist()
```
you will just have to make sure the focus_array_min is sorted